packages/grafana-prometheus/src/querybuilder/components/MetricCombobox.tsx

```diff
@@ -0,0 +1,198 @@
+
+import { PrometheusDatasource } from '../../datasource';
+import { QueryBuilderLabelFilter } from '../shared/types';
+import { PromVisualQuery } from '../types';
+
+import { MetricsModal } from './metrics-modal/MetricsModal';
+
+export interface MetricComboboxProps {
+  metricLookupDisabled: boolean;
+  query: PromVisualQuery;
+  onChange: (query: PromVisualQuery) => void;
+  onGetMetrics: () => Promise<SelectableValue[]>;
+  datasource: PrometheusDatasource;
+  labelsFilters: QueryBuilderLabelFilter[];
+  onBlur?: () => void;
+  variableEditor?: boolean;
+}
+
+export const PROMETHEUS_QUERY_BUILDER_MAX_RESULTS = 1000;
+
+export function MetricCombobox({
+  datasource,
+  query,
+  onChange,
+  onGetMetrics,
+  labelsFilters,
+  metricLookupDisabled,
+  onBlur,
+  variableEditor,
+}: Readonly<MetricComboboxProps>) {
+  const [state, setState] = useState<{
+    metrics?: SelectableValue[];
+    isLoading?: boolean;
+    metricsModalOpen?: boolean;
+    initialMetrics?: string[];
+    resultsTruncated?: boolean;
+  }>({});
+
+  const prometheusMetricEncyclopedia = config.featureToggles.prometheusMetricEncyclopedia;
+
+  /**
+   * Reformat the query string and label filters to return all valid results for current query editor state
+   */
+  // const formatKeyValueStringsForLabelValuesQuery = (
+  //   query: string,
+  //   labelsFilters?: QueryBuilderLabelFilter[]
+  // ): string => {
+  //   const queryString = regexifyLabelValuesQueryString(query);
+
+  //   return formatPrometheusLabelFiltersToString(queryString, labelsFilters);
+  // };
+
+  /**
+   * Gets label_values response from prometheus API for current autocomplete query string and any existing labels filters
+   */
+  // const getMetricLabels = (query: string) => {
+  //   // Since some customers can have millions of metrics, whenever the user changes the autocomplete text we want to call the backend and request all metrics that match the current query string
+  //   const results = datasource.metricFindQuery(formatKeyValueStringsForLabelValuesQuery(query, labelsFilters));
+  //   return results.then((results) => {
+  //     const resultsLength = results.length;
+  //     truncateResult(results);
+
+  //     if (resultsLength > results.length) {
+  //       setState({ ...state, resultsTruncated: true });
+  //     } else {
+  //       setState({ ...state, resultsTruncated: false });
+  //     }
+
+  //     const resultsOptions = results.map((result) => {
+  //       return {
+  //         label: result.text,
+  //         value: result.text,
+  //       };
+  //     });
+
+  //     if (prometheusMetricEncyclopedia) {
+  //       return [...metricsModalOption, ...resultsOptions];
+  //     } else {
+  //       return resultsOptions;
+  //     }
+  //   });
+  // };
+
+  const [comboboxOptions, setCombobboxOptions] = useState<ComboboxOption[]>([]);
+
+  // Load initial options for the dropdown immediately on mount
```

Comment:
It might be worth checking the impact of this when having multiple selects, from a performance and network perspective.

Reply:
You're right - this should have been a TODO. At the moment we don't have a callback to make this lazy.